### PR TITLE
Added bastion_host to all.yaml on AWS

### LINKED ISSUE #267: Add Bastion Host option to the AWS cloud_infrastructure skeleton
Add Bastion Host option to the AWS cloud_infrastructure skeleton

### PR DESCRIPTION
**Pull request by gmontalvoy**
#### What does this PR do?
Added bastion_host on AWS inventories as requested on #267 

#### How should this be manually tested?
Changes added to on casl-ansible/inventory/sample.aws.example.com.d/inventory/group_vars/all.yml

#### Is there a relevant Issue open for this?
Resolve #267 

#### Who would you like to review this?
cc: @redhat-cop/casl @makentenza

### CODE REVIEW

**Comment by gmontalvoy** [comment]
Kindly request for a second review.
Thanks @makentenza

File: roles/manage-aws-infra/README.md [hunk 1]
```diff
@@ -65,6 +65,14 @@ cloud_infrastructure: **mandatory**
      root_volume_size: **defaulted**
      docker_volume_size: **defaulted**
      gluster_volume_size: **defaulted**
+	 bastionnode:  **not mandatory**
```
Comment (makentenza) [inline roles/manage-aws-infra/README.md]: Please review the MD format here as is not legible

File: roles/manage-aws-infra/defaults/main.yml [hunk 2]
```diff
@@ -132,13 +143,15 @@ etcd_root_volume_size: "{{ cloud_infrastructure.etcdnodes.root_volume_size  }}"
...
 cns_node_root_volume_size: "{{ cloud_infrastructure.cnsnodes.root_volume_size }}"
+bastion_node_root_volume_size: "{{ cloud_infrastructure.bastion.root_volume_size }}"
 
 ## Feed root volume device sizes for each instance from `cloud_infrastructure` list on the inventory
 master_docker_volume_size: "{{ cloud_infrastructure.masters.docker_volume_size }}"
 etcd_docker_volume_size: "{{ cloud_infrastructure.etcdnodes.docker_volume_size  }}"
 infra_node_docker_volume_size: "{{ cloud_infrastructure.infranodes.docker_volume_size }}"
 app_node_docker_volume_size: "{{ cloud_infrastructure.appnodes.docker_volume_size }}"
 cns_node_docker_volume_size: "{{ cloud_infrastructure.cnsnodes.docker_volume_size }}"
+bastion_node_docker_volume_size: "{{ cloud_infrastructure.bastion.docker_volume_size }}"
```
Comment (makentenza) [inline roles/manage-aws-infra/defaults/main.yml]: The inventory skeleton is not adding any volume for Docker in the Bastion Nodes

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 3]
```diff
@@ -27,6 +27,10 @@ Parameters:
     Default: "m4.xlarge"
     Description: "The instance type to be used"
     Type: "String"
+  BastionInstanceType:
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Indentation is wrong for 'BastionInstanceType' entries

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 4]
```diff
@@ -232,6 +236,25 @@ Resources:
       Tags:
       - Key: 'Name'
         Value: '{{ env_id }}'
+  bastionsg:
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Indentation is wrong for 'bastionsg' block entries

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 5]
```diff
@@ -428,6 +451,46 @@ Resources:
             DeleteOnTermination: "true"
             VolumeSize: "{{ etcd_docker_volume_size }}"
 {% endfor %}
+{% for idx in range(0, aws_num_bastion | int) %}
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Please review this loop ({% for idx in range(0, aws_num_bastion | int) %}) as you are adding bastion instances here but there are references to the etcd instance type variables. I think you forgot to adapt this block after copy-pasting from previous block

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 6]
```diff
@@ -714,6 +780,13 @@ Resources:
         ResourceRecords:
          - !GetAtt ETCD{{ '%02d' % idx }}.PrivateDnsName
       {% endfor -%}
+      {% for idx in range(0, aws_num_bastion_nodes | int) -%}
+      - Name: "bastion-{{ idx }}.{{ env_id }}.{{ dns_domain }}."
+        Type: CNAME
+        TTL: 300
+        ResourceRecords:
+         - !GetAtt Node{{ '%02d' % idx }}.PrivateDnsName
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: You are getting the incorrect resource here

File: roles/manage-aws-infra/defaults/main.yml (hunk 2 above)
Comment (gmontalvoy) [reply roles/manage-aws-infra/defaults/main.yml]: Fixed

**Comment by gmontalvoy** [comment]
Hi Mak, thank you for review, fixed as requested.

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 7]
```diff
@@ -53,7 +57,7 @@ Parameters:
     Type: "String"
 Resources:
 {# Figure out the node types with more zones to create all the required SubNets per AZ #}
-  {% if (aws_masters_zones.__len__() >= aws_etcd_zones.__len__()) and (aws_masters_zones.__len__() >= aws_appnodes_zones.__len__()) and (aws_masters_zones.__len__() >= aws_infranodes_zones.__len__()) and (aws_masters_zones.__len__() >= aws_cnsnodes_zones.__len__()) -%}
+  {% if (aws_masters_zones.__len__() >= aws_etcd_zones.__len__()) and (aws_masters_zones.__len__() >= aws_bastion_zones.__len__()) and (aws_masters_zones.__len__() >= aws_appnodes_zones.__len__()) and (aws_masters_zones.__len__() >= aws_infranodes_zones.__len__()) and (aws_masters_zones.__len__() >= aws_cnsnodes_zones.__len__()) -%}
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Let's consider the number of Zones for Bastion will never be higher than the number of zones for any other Node type, so please remove the conditional for `(aws_masters_zones.__len__() >= aws_bastion_zones.__len__())`

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 8]
```diff
@@ -747,6 +815,9 @@ Resources:
     {% for idx in range(0, aws_num_etcd | int) -%}
     - ETCD{{ '%02d' % idx }}
     {% endfor -%}
+    {% for idx in range(0, aws_num_bastion | int) -%}
+    - BAST{{ '%02d' % idx }}
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Check that you are declaring the bastion objects as 'Bastion{{ '%02d' % idx }}', not 'BAST{{ '%02d' % idx }}'. Please change this.

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 9]
```diff
@@ -785,3 +856,10 @@ Resources:
         ResourceRecords:
         - !GetAtt ETCD{{ '%02d' % idx }}.PublicDnsName
       {% endfor -%}
+      {% for idx in range(0, aws_num_bastion| int) -%}
+      - Name: "bastion-{{ idx }}.{{ env_id }}.{{ dns_domain }}."
+        Type: CNAME
+        TTL: 300
+        ResourceRecords:
+        - !GetAtt BAST{{ '%02d' % idx }}.PublicDnsName
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Check that you are declaring the bastion objects as 'Bastion{{ '%02d' % idx }}', not 'BAST{{ '%02d' % idx }}'. Please change this.

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 10]
```diff
@@ -676,6 +734,9 @@ Resources:
     {% for idx in range(0, aws_num_etcd | int) -%}
     - ETCD{{ '%02d' % idx }}
     {% endfor -%}
+    {% for idx in range(0, aws_num_bastion | int) -%}
+    - BAST{{ '%02d' % idx }}
```
Comment (makentenza) [inline roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Check that you are declaring the bastion objects as 'Bastion{{ '%02d' % idx }}', not 'BAST{{ '%02d' % idx }}'. Please change this.

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 11]
```diff
@@ -714,6 +775,13 @@ Resources:
         ResourceRecords:
          - !GetAtt ETCD{{ '%02d' % idx }}.PrivateDnsName
       {% endfor -%}
+      {% for idx in range(0, aws_num_bastion_nodes | int) -%}
```
[the remark this answers is not in the document]
Comment (gmontalvoy) [reply roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Changed, I didn't found any other aws_num_bastion_nodes parameter inside that j2 template.

File: roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2 [hunk 12]
```diff
@@ -428,6 +451,41 @@ Resources:
...
+        - !GetAtt sshsg.GroupId
+      Tags:
+        - Key: 'Name'
+          Value: '{{ aws_bastion_name_prefix }}-{{ idx }}.{{ env_id }}.{{ dns_domain }}'
+        - Key: 'ocp_role'
+          Value: 'bastion'
+        - Key: 'env_id'
+          Value: '{{ env_id }}'
+        - Key: 'group'
+          Value: '{{ group_bastion_nodes_tag }}'
```
[the remark this answers is not in the document]
Comment (gmontalvoy) [reply roles/manage-aws-infra/templates/casl-aws-cformation.yml.j2]: Added tag and sgroup as this SG will be maintained for extended purposes as indicated in https://github.com/redhat-cop/casl-ansible/pull/309#discussion_r244299637

**Comment by makentenza** [comment]
@oybed @tylerauerbeck can you guys give this a try to double check is working fine? I did the validation already and worked for me but another look would be great before merging.

**Comment by makentenza** [comment]
@bvkin could you please give it a try to this? Ping me if you need any help.

**Comment by bvkin** [comment]
@makentenza I should have access to within the next day or so. Once I do, I am planning on testing :)

**Comment by bvkin** [comment]
@makentenza @gmontalvoy So I'm currently in the process of testing this. One thing I noticed is that the [galaxy dependencies](https://github.com/gmontalvoy/casl-ansible/blob/aws_bastion_host/casl-requirements.yml) from the pr branch are a bit behind. Did you guys test this with the versions in the casl-requirements.yml from that branch?

**Comment by makentenza** [comment]
tested with the latest tag available in January, but it should be good to go with latest available tag now, give it a try @bvkin and let me know if you find any issue

**Comment by gmontalvoy** [comment]
bump, ¿news?

**Comment by makentenza** [comment]
@gmontalvoy I know this is going quite slow but @bvkin who was doing the final review has been fully allocated to other projects and we didn't have the time to pair on this so far. Let's see if we can get this merged soon

**Comment by bvkin** [comment]
Sorry for the delay, finally got a chance to test this fully. Works for me. The only thing is that the CloudFormation template did not include a section for env_id. This means that the bastion does not get updated after it's provisioned with things like subscription registration. That can be a follow up though if needed.

**Comment by makentenza** [comment]
Thanks for reviewing it @bvkin. As discussed offline env_id for Bastion is something we can add later if required as is not really mandatory for the OCP deployment purpose. Merging it now.

**Pull request merged**